### PR TITLE
Basic support for strace

### PR DESCRIPTION
**Pull request by saagarjha**
It works on simple programs I tried, like `dmesg`.

### CODE REVIEW

File: kernel/ptrace.c [hunk 1]
```diff
@@ -177,7 +183,8 @@ dword_t sys_ptrace(dword_t request, dword_t pid, addr_t addr, dword_t data) {
             if (!child) return _EPERM;
 
             struct user_regs_struct_ user_regs_ = {};
-            get_user_regs(&child->cpu, &user_regs_);
+            get_user_regs_and_syscall(child, &user_regs_);
+            user_regs_.orig_eax = child->ptrace.syscall;
```
Comment (tbodt) [inline kernel/ptrace.c:187]: I believe this line is not needed
```suggestion
```

File: kernel/ptrace.c [hunk 2]
```diff
@@ -44,6 +44,12 @@ static void get_user_regs(struct cpu_state *cpu, struct user_regs_struct_ *user_
 //  user_regs_->xss = cpu->xss;
 }
 
+// Ensure stopped, ptrace locked, etc. before calling this
+static void get_user_regs_and_syscall(struct task *task, struct user_regs_struct_ *user_regs_) {
+    get_user_regs(&task->cpu, user_regs_);
+    user_regs_->orig_eax = task->ptrace.syscall;
```
Comment (tbodt) [inline kernel/ptrace.c:50]: Should this be folded into get_user_regs? I see no other users of get_user_regs.

File: kernel/task.h [hunk 3]
```diff
@@ -54,9 +54,13 @@ struct task {
 
         bool traced;
         bool stopped;
+        bool sysgood;
+        bool stop_at_syscall;
+        bool syscall_stopped;
```
Comment (yezz123) [inline kernel/task.h:59]: Should that be called as Boolean form ?

Comment (saagarjha) [reply kernel/task.h:59]: I'm not sure what you mean?

**Pull request closed without merging**